src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp

```diff
@@ -0,0 +1,901 @@
+    // other clients liveliness is provided through server's PDP discovery data
+
+    // check if the DATA msg is relayed by another server
+    bool do_lease = participant_data.m_guid.guidPrefix == change.writerGUID.guidPrefix;
+
+    if (!do_lease)
+    {
+        // if not a client verify this participant is a server
+        for (auto & svr : mp_builtin->m_DiscoveryServers)
+        {
+            if (svr.guidPrefix == participant_data.m_guid.guidPrefix)
+            {
+                do_lease = true;
+            }
+        }
+    }
+
+    ParticipantProxyData * pdata = new ParticipantProxyData(participant_data);
+    pdata->isAlive = true;
+
+    if (do_lease)
+    {
+        pdata->mp_leaseDurationTimer = new RemoteParticipantLeaseDuration(this,
+            pdata,
+            TimeConv::Duration_t2MilliSecondsDouble(pdata->m_leaseDuration));
+        pdata->mp_leaseDurationTimer->restart_timer();
+    }
+    else
+    {
+        pdata->mp_leaseDurationTimer = nullptr;
+    }
+
+    m_participantProxies.push_back(pdata);
+
+    return pdata;
+}
+
+bool PDPServer::createPDPEndpoints()
+{
+    logInfo(RTPS_PDP, "Beginning PDPServer Endpoints creation");
+
+    HistoryAttributes hatt;
+    hatt.payloadMaxSize = DISCOVERY_PARTICIPANT_DATA_MAX_SIZE;
+    hatt.initialReservedCaches = pdp_initial_reserved_caches;
+    hatt.memoryPolicy = mp_builtin->m_att.readerHistoryMemoryPolicy;
+    mp_PDPReaderHistory = new ReaderHistory(hatt);
+
+    ReaderAttributes ratt;
+    ratt.expectsInlineQos = false;
+    ratt.endpoint.endpointKind = READER;
+    ratt.endpoint.multicastLocatorList = mp_builtin->m_metatrafficMulticastLocatorList;
+    ratt.endpoint.unicastLocatorList = mp_builtin->m_metatrafficUnicastLocatorList;
+    ratt.endpoint.topicKind = WITH_KEY;
+    ratt.endpoint.durabilityKind = TRANSIENT_LOCAL;
+    ratt.endpoint.reliabilityKind = RELIABLE;
+    ratt.times.heartbeatResponseDelay = pdp_heartbeat_response_delay;
+
+    mp_listener = new PDPServerListener(this);
+
+    if (mp_RTPSParticipant->createReader(&mp_PDPReader, ratt, mp_PDPReaderHistory,
+        mp_listener, c_EntityId_SPDPReader, true, false))
+    {
+        // enable unknown clients to reach this reader
+        mp_PDPReader->enableMessagesFromUnkownWriters(true);
+
+        for (auto it = mp_builtin->m_DiscoveryServers.begin(); it != mp_builtin->m_DiscoveryServers.end(); ++it)
+        {
+            RemoteWriterAttributes rwatt;
+
+            rwatt.guid = it->GetPDPWriter();
+            rwatt.endpoint.multicastLocatorList.push_back(it->metatrafficMulticastLocatorList);
+            rwatt.endpoint.unicastLocatorList.push_back(it->metatrafficUnicastLocatorList);
+            rwatt.endpoint.topicKind = WITH_KEY;
+            rwatt.endpoint.durabilityKind = _durability; // Server Information must be persistent
+            rwatt.endpoint.reliabilityKind = RELIABLE;
+
+            // TODO: remove the join when Reader and Writer match functions are updated
```

Comment:
Right now StatefulReader::matched_writer_add(...) is using remoteLocatorList to create the sender resources (thus I must muster all locators to that list). In the future this method would be updated and remoteLocatorList use deprecated.

#5755 - Update PDPServer::assignRemoteEndpoints to avoid EndpointAttributes::remoteLocatorList use